[GStreamer] Crash in CachedResourceStreamingClient::dataReceived

philn opened the pull request:
#### 6a9d7a164705ebaf78a158be71f97a5f520f9655
<pre>
[GStreamer] Crash in CachedResourceStreamingClient::dataReceived
<a href="https://bugs.webkit.org/show_bug.cgi?id=266973">https://bugs.webkit.org/show_bug.cgi?id=266973</a>

Reviewed by Michael Catanzaro, Carlos Garcia Campos and Xabier Rodriguez-Calvar.

Store the WebKitWebSrc pointer as a weak reference in the CachedResourceStreamingClient. Using a
const raw pointer was not sufficient, because nothing prevented other code to dispose the element,
thus making the pointer point to garbage. We can&apos;t use a strong reference because it would introduce
a reference cycle between WebKitWebSrc and the CachedResourceStreamingClient.

* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(webKitWebSrcMakeRequest):
(CachedResourceStreamingClient::CachedResourceStreamingClient):
(CachedResourceStreamingClient::~CachedResourceStreamingClient):
(CachedResourceStreamingClient::checkUpdateBlocksize):
(CachedResourceStreamingClient::responseReceived):
(CachedResourceStreamingClient::dataReceived):
(CachedResourceStreamingClient::accessControlCheckFailed):
(CachedResourceStreamingClient::loadFailed):
(CachedResourceStreamingClient::loadFinished):

Canonical link: <a href="https://commits.webkit.org/272807@main">https://commits.webkit.org/272807@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f945af303527332a58844c460d013e149531752c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32845 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11599 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34715 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35463 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29669 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33777 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13945 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8778 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29119 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33295 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9782 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29330 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8516 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8663 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29286 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36796 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/28105 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29840 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29691 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34784 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/32850 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8780 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6743 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32652 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10470 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/39317 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7677 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9389 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8282 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9402 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -942,32 +942,38 @@ bool webKitSrcPassedCORSAccessCheck(WebKitWebSrc* src)
     ASSERT(isMainThread());
-    WebKitWebSrc* src = WEBKIT_WEB_SRC(m_src);
-    GstBaseSrc* baseSrc = GST_BASE_SRC_CAST(src);
+    GRefPtr<WebKitWebSrc> src(WEBKIT_WEB_SRC_CAST(g_weak_ref_get(&m_src)));
```
g_weak_ref_get returns a reference, so you need to adopt it. We already have GWeakPtr in WTF, but I guess we need thread safety here, so maybe we can add GThreadSafeWeakPtr to wrap GWeakRef and make it more difficult (or impossible) to make this kind of mistakes.

philn replied:
TIL `WTF::GWeakPtr`! ThreadSafe as in check the ptr would be used from the same thread it was created from?

mcatanzaro replied:
I agree we should add `GThreadSafeWeakPtr`.

> ThreadSafe as in check the ptr would be used from the same thread it was created from?

No. (We should probably add an ASSERT() to the existing GWeakPtr to guarantee that, though. WTF::WeakPtr already has a similar check.)

GThreadSafeWeakPtr would be a weak pointer that guarantees you either atomically acquire a strong ref (preventing the object from being destroyed on another thread) or fail to do so (if the object is already being destroyed). It would need to be implemented using GWeakRef (as suggested by Carlos) directly rather than g_object_add_weak_pointer(). `GThreadSafeWeakPtr::get` would need to return a `GRefPtr` rather than a raw pointer.

mcatanzaro replied:
Also we would want to add a test to Tools/TestWebKitAPI/Tests/WTF/glib/GWeakPtr.cpp.

I don't think GThreadSafeWeakPtr would be too hard to create, but I also wouldn't want to make creating it a condition for fixing this bug.

mcatanzaro replied:
Oh, I see you've already done exactly this, and I just wasted my time by explaining what to do. Nice.

mcatanzaro replied:
> Also we would want to add a test to Tools/TestWebKitAPI/Tests/WTF/glib/GWeakPtr.cpp.

A quick test here would be nice to have, though.

philn replied:
OK, I'll write a test tomorrow!

carlosgcampos commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -991,15 +997,20 @@ void CachedResourceStreamingClient::checkUpdateBlocksize(unsigned bytesRead)
 {
     ASSERT(isMainThread());
-    WebKitWebSrc* src = WEBKIT_WEB_SRC(m_src);
+    GRefPtr<WebKitWebSrc> src(WEBKIT_WEB_SRC_CAST(g_weak_ref_get(&m_src)));
```
Same here.

carlosgcampos commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -1111,7 +1122,10 @@ void CachedResourceStreamingClient::redirectReceived(PlatformMediaResource&, Res
 {
     ASSERT(isMainThread());
-    WebKitWebSrc* src = WEBKIT_WEB_SRC(m_src);
+    GRefPtr<WebKitWebSrc> src(WEBKIT_WEB_SRC_CAST(g_weak_ref_get(&m_src)));
```
And here

carlosgcampos commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -1124,55 +1138,61 @@ void CachedResourceStreamingClient::dataReceived(PlatformMediaResource&, const S
 {
     ASSERT(isMainThread());
-    WebKitWebSrc* src = WEBKIT_WEB_SRC(m_src);
+    GRefPtr<WebKitWebSrc> src(WEBKIT_WEB_SRC_CAST(g_weak_ref_get(&m_src)));
```
And here, and well everywhere.

mcatanzaro reviewed (approved):
GThreadSafeWeakPtr and its tests look good.

The changes to the multimedia code look fine to me, but you might want a multimedia reviewer to look too.

carlosgcampos commented on Source/WTF/wtf/glib/GThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,89 @@
+    {
+        auto strongRef = get();
+        return !strongRef;
+    }
```
I'm not sure this is a good idea in this particular case because get is not cheap, so doing something like:

```
if (foo)
    bar(foo.get());
```

means we take and release the mutex twice and create a strong reference twice.

philn replied:
Yeah I had the same thoughts... that's why i didn't use this operator in the patch. I can remove it maybe?

mcatanzaro replied:
Eh, it's debatable. I think the normal pattern here will be to get a strong ref and then do something with it if you succeed. Checking whether the weak pointer is still valid and then doing something would be weird because it could be invalidated between your check and your doing something.  So, maybe it would be best to not provide any operator!

carlosgcampos replied:
Yes, I think it's better to not provide operator!

carlosgcampos reviewed (approved):
Looks great. Thank you!

philn committed:
[GStreamer] Crash in CachedResourceStreamingClient::dataReceived
https://bugs.webkit.org/show_bug.cgi?id=266973

Reviewed by Michael Catanzaro, Carlos Garcia Campos and Xabier Rodriguez-Calvar.

Store the WebKitWebSrc pointer as a weak reference in the CachedResourceStreamingClient. Using a
const raw pointer was not sufficient, because nothing prevented other code to dispose the element,
thus making the pointer point to garbage. We can't use a strong reference because it would introduce
a reference cycle between WebKitWebSrc and the CachedResourceStreamingClient.

* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(webKitWebSrcMakeRequest):
(CachedResourceStreamingClient::CachedResourceStreamingClient):
(CachedResourceStreamingClient::~CachedResourceStreamingClient):
(CachedResourceStreamingClient::checkUpdateBlocksize):
(CachedResourceStreamingClient::responseReceived):
(CachedResourceStreamingClient::dataReceived):
(CachedResourceStreamingClient::accessControlCheckFailed):
(CachedResourceStreamingClient::loadFailed):
(CachedResourceStreamingClient::loadFinished):

Canonical link: https://commits.webkit.org/272807@main
